flashinfer/comm/allreduce.py

```diff
@@ -0,0 +1,746 @@
+
+    Raises:
+        BackendSupportedError: If no suitable backend available for the configuration
+        ValueError: If problem size not supported for the specified backend
+
+    Examples:
+        >>> # Auto-select best backend based on topology
+        >>> workspace = create_allreduce_fusion_workspace(
+        ...     backend="auto",
+        ...     world_size=8,
+        ...     rank=0,
+        ...     max_token_num=2048,
+        ...     hidden_dim=4096,
+        ...     dtype=torch.bfloat16,
+        ...     topology="single_node"
+        ... )
+        >>> print(workspace.backend)  # "trtllm"
+        >>> print(workspace.get_workspace_capacity())  # 8388608 elements
+
+        >>> # Check if workspace can handle different problem sizes
+        >>> workspace.is_sufficient_for(1024, 4096, 8, torch.bfloat16)  # True
+        >>> workspace.is_sufficient_for(4096, 2048, 8, torch.bfloat16)  # True (same total)
+
+        >>> # Explicit backend selection
+        >>> workspace = create_allreduce_fusion_workspace(
+        ...     backend="mnnvl",
+        ...     world_size=16,
+        ...     rank=0,
+        ...     max_token_num=2048,
+        ...     hidden_dim=4096,
+        ...     dtype=torch.bfloat16,
+        ...     topology="multi_node"
+        ... )
+        >>> print(workspace.backend)  # "mnnvl"
+    """
+    # Decorator has validated backend - now create workspace
+    # If backend="auto", decorator has selected the best one and stored it
+
+    # Get actual backend (decorator resolved "auto" to concrete backend)
+    if backend == "auto":
+        # Decorator stored the selected backend in suitable_auto_backends
+        actual_backend = create_allreduce_fusion_workspace.suitable_auto_backends[0]
+    else:
+        actual_backend = backend
+
+    # Create workspace for selected backend using workspace constructors
+    if actual_backend == "trtllm":
+        return TRTLLMAllReduceFusionWorkspace(
+            tp_size=world_size,
+            tp_rank=rank,
+            max_token_num=max_token_num,
+            hidden_dim=hidden_dim,
+            dtype=dtype,
+            process_group=process_group,
+        )
+
+    elif actual_backend == "mnnvl":
+        return MNNVLAllReduceFusionWorkspace(
+            world_size=world_size,
+            rank=rank,
+            max_token_num=max_token_num,
+            hidden_dim=hidden_dim,
+            dtype=dtype,
+            **backend_kwargs,
+        )
+    else:
+        raise RuntimeError(f"Unknown backend: {actual_backend}")
+
+
+# ============================================================================
+# MAIN API - NO backend parameter, infers from workspace type
+# ============================================================================
+
+
+def allreduce_fusion(
+    input: torch.Tensor,
+    workspace: AllReduceFusionWorkspace,
+    pattern: int,
+    launch_with_pdl: bool = False,
```

Comment:
Why the advantage to give pdl control to the user?

Reply:
We have been doing this for all our APIs, but I am not sure why. Maybe because not all archs support it?